tests/integration/test_l1_read_request.py

```diff
@@ -0,0 +1,283 @@
+        - Read lines to fill the cache (cache misses)
+        - Read line to cause an eviction (cache miss)
+        """
+        # Base addr that returns a NOHIT snoop response from other caches
+        address = 0x00000002  # Start with set 0, offset with NOHIT response
+        increment = 0x100000  # 1 MiB increment to change the tag
+
+        # Fill the cache
+        for _i in range(16):
+            address += increment
+            handle_event(self.cache, self.trace_event, address)
+            self.check_line_state(address, MESIState.EXCLUSIVE)
+
+        # One more read to trigger an eviction of a clean line
+        handle_event(self.cache, self.trace_event, address + increment)
+
+        # FIXME: Enable this one we've updated the code to send this message
+        # Check for however many sendline events we need (likely 17)
+        # self.mock_logger.assert_called_with(
+        #     LogLevel.NORMAL, Regex(r".*L2.*sendline.*")
+        # )
+
+        # Check for one READ bus operation per cache miss
+        self.assert_log_called_with_count(LogLevel.NORMAL, r".*busop.*(read|1).*", 17)
+
+        # Assertions for statistics
+        self.assertEqual(self.cache.statistics.cache_reads, 17)
+        self.assertEqual(self.cache.statistics.cache_writes, 0)
+        self.assertEqual(self.cache.statistics.cache_hits, 0)
+        self.assertEqual(self.cache.statistics.cache_misses, 17)
+
+    def test_dirty_eviction(self):
+        """
+        Test reading a line that causes an eviction of a dirty line.
+        Other caches return a NOHIT snoop response.
+
+        Requires the following sequence:
+        - Write lines to fill the cache (cache misses)
+        - Read line to cause an eviction (cache miss)
+        """
+        # Base addr that returns a NOHIT snoop response from other caches
+        address = 0x00000002  # Start with set 0, offset with NOHIT response
+        increment = 0x100000  # 1 MiB increment to change the tag
+
+        # Fill the cache
+        for _i in range(16):
+            address += increment
+            handle_event(self.cache, 1, address)
+            self.check_line_state(address, MESIState.MODIFIED)
+
+        # One more read to trigger an eviction of a dirty line
+        handle_event(self.cache, self.trace_event, address + increment)
+
+        # FIXME: Enable this one we've updated the code to send this message
+        # Check for however many sendline events we need (likely 17)
+        # self.mock_logger.assert_called_with(
+        #     LogLevel.NORMAL, Regex(r".*L2.*sendline.*")
+        # )
+
+        # Check for one READ bus operation per cache miss
+        self.assert_log_called_with_count(LogLevel.NORMAL, r".*busop.*(rwim|4).*", 16)
+        # FIXME: Check functionality of code for this case - the test found 4
+        # matching calls to this bus op read event. Is that correct?
+        # self.assert_log_called_with_count(LogLevel.NORMAL, r".*busop.*(read|1).*", 1)
```

Comment:
I *believe* I have this test set up such that I'd only expect rwim bus ops for the looped write events, and not read events, but I could use another set of eyes to confirm @reecewayt. And if the test looks correct, then we need to consider whether the code is behaving correctly.

Reply:
@nkanderson The test looks correct and I traced the calls this would make an I believe the MESI controller has the correct logic. 
```python 
#this would be the logic trigger in the test's loop
elif current_state == MESIState.INVALID and is_processor_write:
            bus_operation(BusOp.RWIM, address)  # Bus Read for ownership
            return MESIState.MODIFIED
```

With that said maybe there is  something going on with mocked logger, but that would be weird considering there's no problems with the `test_clean_eviction` test.

Reply:
> For debugging stuff like this what do you think about redirecting the DEBUG log levels in the mocked class to standard out?

I think the standard error output will just be written to the console when running the program on the command line. Is that what you mean?